haystack/preview/components/reader.py

```diff
@@ -0,0 +1,150 @@
+from haystack.lazy_imports import LazyImport
+
+with LazyImport(message="Run 'pip install farm-haystack[inference]'") as torch_and_transformers_import:
+    from transformers import AutoModelForQuestionAnswering
+    from tokenizers import Tokenizer, Encoding
+    import torch
+
+
+@component
+class ExtractiveReader:
+    def __init__(
+        self, reader: Union[Path, str], device: Optional[str] = None, max_seq_length: int = 384, top_k: int = 10
+    ) -> None:
+        self.reader = reader
+        self.device = device
+        self.loaded = False
+        self.max_seq_length = max_seq_length
+        self.top_k = top_k
+
+    def warm_up(self):
+        if not self.loaded:
+            torch_and_transformers_import.check()
+            if torch.cuda.is_available():
+                self.device = "cuda:0"
+            else:
+                self.device = "cpu:0"
+            self.model = AutoModelForQuestionAnswering.from_pretrained(self.reader).to(self.device)
+            self.tokenizer = Tokenizer.from_pretrained(self.reader)
+
+    def _flatten(self, queries: List[str], documents: List[List[Document]]) -> Tuple[List[str], List[str]]:
+        flattened_queries = [query for documents_, query in zip(documents, queries) for _ in documents_]
+        flattened_documents = [document.content for documents_ in documents for document in documents_]
+        return flattened_queries, flattened_documents
+
+    def _preprocess(
+        self, queries: List[str], documents: List[str], max_seq_length: int
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, List[Encoding]]:
+        self.tokenizer.enable_padding(length=max_seq_length)
+        self.tokenizer.enable_truncation(max_seq_length)
+
+        encodings = self.tokenizer.encode_batch(list(zip(queries, documents)))
+
+        input_ids = torch.tensor([encoding.ids for encoding in encodings]).to(self.device)
+        attention_mask = torch.tensor([encoding.attention_mask for encoding in encodings]).to(self.device)
+        sequence_ids = torch.tensor(
+            [[id_ if id_ is not None else -1 for id_ in encoding.sequence_ids] for encoding in encodings]
+        ).to(self.device)
+
+        return input_ids, attention_mask, sequence_ids, encodings
+
+    def _postprocess(
+        self,
+        output,
+        sequence_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        top_k: int,
+        encodings: List[Encoding],
+        max_seq_length: int,
+    ) -> Tuple[List[List[int]], List[List[int]], List[List[float]]]:
+        start = output.start_logits
+        end = output.end_logits
+        mask = sequence_ids == 1
+        mask[..., 0] = True  # type: ignore # mypy doesn't understand broadcasting
+        mask = torch.logical_and(mask, attention_mask == 1)  # type: ignore # see above
+        start = torch.where(mask, start, -torch.inf)  # type: ignore # see above
+        end = torch.where(mask, end, -torch.inf)  # type: ignore # see above
```

Comment:
Mypy doesn't like this, but I think it's fine.

Reply:
What is the error being raised?

Reply:
The errors are:
```
haystack/preview/components/reader.py:67: error: Unsupported target for indexed assignment ("bool")  [index]
haystack/preview/components/reader.py:68: error: Incompatible types in assignment (expression has type "Tensor", variable has type "bool")  [assignment]
haystack/preview/components/reader.py:68: error: Argument 1 to "logical_and" has incompatible type "bool"; expected "Tensor"  [arg-type]
haystack/preview/components/reader.py:68: error: Argument 2 to "logical_and" has incompatible type "bool"; expected "Tensor"  [arg-type]
haystack/preview/components/reader.py:69: error: No overload variant of "where" matches argument types "bool", "Any", "float"  [call-overload]
haystack/preview/components/reader.py:69: note: Possible overload variants:
haystack/preview/components/reader.py:69: note:     def where(condition: Tensor) -> List[Tensor]
haystack/preview/components/reader.py:69: note:     def where(condition: Tensor, input: Tensor, other: Tensor, *, out: Optional[Tensor] = ...) -> Tensor
haystack/preview/components/reader.py:69: note:     def where(condition: Tensor, self: Union[int, float, bool], other: Tensor) -> Tensor
haystack/preview/components/reader.py:69: note:     def where(condition: Tensor, input: Tensor, other: Union[int, float, bool]) -> Tensor
haystack/preview/components/reader.py:69: note:     def where(condition: Tensor, self: Union[int, float, bool], other: Union[int, float, bool]) -> Tensor
```
I think it doesn't understand the assignment in line 66 and then gets the type of mask wrong. It also doesn't understand that `attention_mask == 1` returns a tensor and not a bool. This causes it not to understand line 67 and just to type mask as `Any` which causes the errors in the following lines.

Reply:
Maybe explicit type hinting can fix it, something like:

```python
mask: torch.Tensor = sequence_ids == 1
```

Reply:
I am not sure why, but it seems to work now even without adding that hint.